### DISCUSSION TITLE
set_cmd discrepancy with more than one parameter in my driver

### DISCUSSION THREAD

**Discussion by cmordi** [microsoft/Qcodes#5607] [General]
Hi, 

I am currently writing a driver for my Ametek 5210 and I have run into a roadblock.  In the manual I am told that the layout for vernier control is G [n1, [n2]] and told that that the command for filter frequency is FF [n1, n2], but when trying to write it in my driver like the two examples below, it doesn't work:
```

class Model5210LockIn(VisaInstrument):
        def __init__(self, name, address, **kwargs):
           super().__init__(name, address, **kwargs)

           self.add_parameter('filter_frequency',
                       get_cmd='FF?',
                       set_cmd='FF {} {}', 
                       unit='Hz',
                       docstring='Controls the filter filtering band.')

           self.add_parameter('vernier_gain',
                       get_cmd='G?',
                       set_cmd= 'G {} {}',
                       vals = vals.Numbers(min_value=0, max_value=255),
                       docstring='Sets or reads the status and gain of the signal-channel gain vernier control.')
```

I know it doesn't work because in my example jupyter notebook I tested it with two different scenarios:

`amtk_lockin.vernier_gain(1, 10)` yields

`TypeError: ParameterBase._wrap_set.<locals>.set_wrapper() takes 1 positional argument but 2 were given`

and
`amtk_lockin.vernier_gain(1)`  yields:

`IndexError: ('Replacement index 1 out of range for positional args tuple', 'setting locking_in_vernier_gain to 1')`

Has anyone ever encountered this or had a similar issue? Can't seem to find solutions anywhere.

**Comment by jenshnielsen** [comment]
Hi, The parameters is not really designed to take more than one argument. 

There are two options you could let the parameter take an input as a tuple of the 2 values
such that you would call it with `amtk_lockin.vernier_gain((1, 10))` In that case you would also need to update your validator to take two argumentets. In general, i'ts also good design that there is sympatry between the get and the set command such that you can call set with the output of get. Does `FF?` and `G?` return two values that should be split?  To make this work you may have to write a custom subclass of Parameter that implements `set_raw` and `get_raw` to insert/extract the values from the string. 

Another alternative would be to use a Group with Grouped parameters like this https://qcodes.github.io/Qcodes/api/parameters/index.html#qcodes.parameters.Group 
This makes more sense if you want the user to be able to set the two input values independently as if they were individual parameters. 

However, it's not clear to me from your parameters why they take two arguments so I don't know which solution makes the most sense.

**Reply by cmordi** [reply]
Hi Jens, 

Thanks for the response. You are correct, `FF?` and `G?` and most of my other commands that require two parameters (`n1` and `n2`) should return two values since it should be sending two values to the instrument. The `n1 `and `n2 `parameters are meant to work together, with `n1 `being a parameter (usually 1 or 0) to switch off/on the channel or set/get which quadrant you want to work in if given different quadrants and `n2 `being the parameter that sets/gets specific values and provides more granular control once you have selected whether you want the channel to be on/off or selected your channel, in some cases n2 is optional. 

Based on what you have told me, the `set_raw` and `get_raw ` approach sounds like it is more relevant to what I am trying to achieve. I may be wrong though
